ethers/signer.nim

```diff
@@ -26,10 +26,12 @@ proc raiseEstimateGasError(
     parent: parent)
   raise e
 
-method provider*(signer: Signer): Provider {.base, gcsafe.} =
+method provider*(signer: Signer):
+    Provider {.base, gcsafe, raises: [CatchableError].} =
```

Comment:
Hmm... I would not blanked raise `CatchableError`, in fact I would set this to `raises: []` and only list those exceptions that are intended to propagate. The whole point of this is preventing spurious exceptions from bubbling up, otherwise there is no point in tracking exceptions at all.

Reply:
In fact I would use the `push: raises []` pragma in all the files and only override this per method on an as needed basis.

Reply:
Yeah, I mean, as I was working through this, I was faced with two choices:

  1. do the minimal amount of changes so that the current implementation works with Chronos V4;
  2. actually track the entire exception chains and expose the minimal set of exceptions that all method implementations may throw.

I went with 1 to get there faster and make the changes smaller. But we can definitely do 2.

Reply:
Yeah, I mean, as I was working through this, I was faced with two choices:

  1. do the minimal amount of changes so that the current implementation works with Chronos V4;
  2. actually track the entire exception chains and expose the minimal set of exceptions that all method implementations may throw, which is what you're proposing.

(2) seems like a good idea, but it is actually sort of tricky. Without some feeling as to how the API has to be designed, I could easily replace `CatchableError` with `EthersError` pretty much everywhere, and it would compile without much gain in exception semantics. The actual narrow types for the exceptions that would make sense to expose depends on what we expect clients to want to try to recover from, so that's why I ended up going with the (1) route.

I'll give it a stab anyhow, let's see what comes out of it.

Reply:
Yeah, I understand, this gets messy pretty quickly, but I still think that push pragma with `raises: []` and an explicit `EthersError` where required is a much better than the bare `CatchableError`. Also, I suspect that a lot of issues are stemming from the fact that callbacks aren't annotated with `raises: []`.